packages/flutter_tools/lib/src/device.dart

```diff
@@ -210,103 +275,133 @@ abstract class DeviceManager {
         ...await discoverer.getDiagnostics(),
     ];
   }
+}
 
-  /// Find and return all target [Device]s based upon currently connected
-  /// devices, the current project, and criteria entered by the user on
-  /// the command line.
-  ///
-  /// If no device can be found that meets specified criteria,
-  /// then print an error message and return null.
-  ///
-  /// Returns a list of devices specified by the user.
-  ///
-  /// * If the user specified '-d all', then return all connected devices which
-  /// support the current project, except for fuchsia and web.
-  ///
-  /// * If the user specified a device id, then do nothing as the list is already
-  /// filtered by [getDevices].
-  ///
-  /// * If the user did not specify a device id and there is more than one
-  /// device connected, then filter out unsupported devices and prioritize
-  /// ephemeral devices.
-  ///
-  /// * If [promptUserToChooseDevice] is true, and there are more than one
-  /// device after the aforementioned filters, and the user is connected to a
-  /// terminal, then show a prompt asking the user to choose one.
-  Future<List<Device>> findTargetDevices(
-    FlutterProject? flutterProject, {
-    Duration? timeout,
-  }) async {
-    if (timeout != null) {
-      // Reset the cache with the specified timeout.
-      await refreshAllConnectedDevices(timeout: timeout);
-    }
+/// A class for determining how to filter devices based on if they are supported.
+///
+/// If [mustBeSupportedByFlutter] is true, only devices supported by Flutter
+/// will be included.
+///
+/// If [mustBeSupportedForProject] is true, only devices supported by Flutter
+/// and supported for the provided [flutterProject] will be included.
+///
+/// If [mustBeSupportedForAll] is true, only devices supported by Flutter,
+/// supported for the provided [flutterProject], and supported for
+/// `--device all` will be included.
+class DeviceDiscoverySupportFilter {
+  DeviceDiscoverySupportFilter({
+    required this.flutterProject,
+    this.mustBeSupportedByFlutter = false,
```

Comment:
How about `includeDevicesSupportedByFlutter`, `includeDevicesSupportedByProject`, `includeUnsupportedDevices` or similar.

Reply:
Yeah I had a hard time picking variable names for this. So by default (if you don't set anything) it will include all devices and none will be filtered out. So I thought that using the word `include` was a bit confusing. For example, if everything is already included, what does `includeDevicesSupportedByFlutter` really mean? It really means exclude everything that's not supported by flutter. What do you think of `excludeDevicesNotSupportedByFlutter`, `excludeDevicesNotSupportedByProject`, `excludeDevicesNotSupportedByAll`?